### PR TITLE
Fix case sensitivity in LoadInstrument

### LINKED ISSUE #26816: LoadInstrument requires specific case in filename
Mantid requires us to have the word "_Def" when calling LoadInstrument to determine the instrument name.
If this is lowercase (or upper) we will instead put the whole filename as the instrument name.

This affects both Windows and Linux as it's the case of the filename parameter which matters, not the file system

### PR DESCRIPTION
**Pull request by DavidFair**
**Description of work.**
Fixes case sensitivity in LoadInstrument; if the filename had _Definition that does not follow the aforementioned case it would instead embed the entire filename as the instrument.
**To test:**
The data is available from unit test data and this bug exists on both Windows and Linux

- Change the `path_to_your_mantid_src` to suit your system
- If your on Windows change the forward slashes to back slashes
- If your on Linux you need to also copy LARMOR_Definition.xml -> LARMOR_definition.xml
- Run the script on an unfixed copy of Mantid, note it says LARMOR_definition.xml
```
from __future__ import (absolute_import, division, print_function, unicode_literals)

from mantid.simpleapi import *

import matplotlib.pyplot as plt

import numpy as np
import tempfile

import h5py

path_to_your_mantid_src = r"/home/ieb35538/mantid/"
# Leave definition lowercase!
path_to_inst_def = r"instrument/LARMOR_definition.xml" 

def get_inst_name(path):
    with h5py.File(path, 'r') as file:
        keys = file.keys()
        first_entry = file[keys[0]]
        inst = first_entry["instrument"]
        name_data = inst["name"]
        inst_name = name_data[0].decode("utf-8")
        print(inst_name)

full_path = path_to_your_mantid_src + path_to_inst_def

ws = Load("LARMOR2260")
LoadInstrument(ws,RewriteSpectraMap=True,Filename=full_path)

with tempfile.NamedTemporaryFile() as tmp:
    print(tmp.name)
    SaveNexusProcessed(ws, Filename=tmp.name)
    get_inst_name(tmp.name)
```
- Pull these changes
- Re-run the script, it should produce LARMOR now.

Fixes #26816 . 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/DataHandling/src/LoadInstrument.cpp [hunk 1]
```diff
@@ -150,8 +152,21 @@ void LoadInstrument::exec() {
     const std::string::size_type stripPath = filename.find_last_of("\\/");
     std::string instrumentFile =
         filename.substr(stripPath + 1, filename.size());
+
     // Strip off "_Definition.xml"
-    instname = instrumentFile.substr(0, instrumentFile.find("_Def"));
+    auto definitionRange = boost::ifind_first(instrumentFile, "_Def");
+    if (definitionRange) {
+      // We need to find the position this occurs in the string and we
```
Comment (martyngigg) [inline Framework/DataHandling/src/LoadInstrument.cpp]: Can you use `std::distance` with something like:
```c++
instname = filename.substr(0, std::distance(filename.begin(), range.begin()));
```
?

File: Framework/DataHandling/src/LoadInstrument.cpp [hunk 2]
```diff
@@ -150,8 +152,21 @@ void LoadInstrument::exec() {
...
+      // can't pass begin() to substr. So take a performance pessimism
+      // and create a copy which find is guaranteed to find
+      const std::string foundDefTagCase(definitionRange.begin(),
+                                        definitionRange.end());
+      instname = instrumentFile.substr(0, instrumentFile.find(foundDefTagCase));
+    } else {
+      g_log.warning("The instrument definition filename does not contain "
```
Comment (martyngigg) [inline Framework/DataHandling/src/LoadInstrument.cpp]: Should be more explicit in the warning and specify that we are setting the instrument name to `instrumentFile`?

**Review by martyngigg — CHANGES_REQUESTED** [review]
Fixes a small hole we've had for ~10 years. Minor comment whether we can avoid a copy.

File: Framework/DataHandling/src/LoadInstrument.cpp (hunk 1 above)
Comment (DavidFair) [reply Framework/DataHandling/src/LoadInstrument.cpp]: I knew the STL must have a way to handle the difference between a normal and string iterator, but just couldn't find it. Thanks for that.

**Review by martyngigg — APPROVED** [review]
Looks good :+1:

**Pull request merged**